lake-primitives/src/types/receipts.rs

```diff
@@ -148,113 +149,518 @@ impl Action {
-        gas: Gas,
-        #[serde(with = "dec_format")]
-        deposit: Balance,
-    },
-    Transfer {
-        #[serde(with = "dec_format")]
-        deposit: Balance,
-    },
-    Stake {
-        #[serde(with = "dec_format")]
-        stake: Balance,
-        public_key: PublicKey,
-    },
-    AddKey {
-        public_key: PublicKey,
-        access_key: views::AccessKeyView,
-    },
-    DeleteKey {
-        public_key: PublicKey,
-    },
-    DeleteAccount {
-        beneficiary_id: AccountId,
-    },
-}
-
-impl From<&views::ActionView> for Operation {
-    fn from(action: &views::ActionView) -> Self {
-        match action {
-            &views::ActionView::CreateAccount => Self::CreateAccount,
-            views::ActionView::DeployContract { code } => {
-                Self::DeployContract { code: code.clone() }
-            }
-            views::ActionView::FunctionCall {
-                method_name,
-                args,
-                gas,
-                deposit,
-            } => Self::FunctionCall {
-                method_name: method_name.to_string(),
-                args: args.clone(),
-                gas: *gas,
-                deposit: *deposit,
-            },
-            views::ActionView::Transfer { deposit } => Self::Transfer { deposit: *deposit },
-            views::ActionView::Stake { stake, public_key } => Self::Stake {
-                stake: *stake,
-                public_key: public_key.clone(),
-            },
-            views::ActionView::AddKey {
-                public_key,
-                access_key,
-            } => Self::AddKey {
-                public_key: public_key.clone(),
-                access_key: access_key.clone(),
-            },
-            views::ActionView::DeleteKey { public_key } => Self::DeleteKey {
-                public_key: public_key.clone(),
-            },
-            views::ActionView::DeleteAccount { beneficiary_id } => Self::DeleteAccount {
-                beneficiary_id: beneficiary_id.clone(),
-            },
-            views::ActionView::Delegate {
-                delegate_action,
-                signature,
-            } => todo!(),
+    // Tries to convert a near_primitives::delegate_action::DelegateAction into a vector of Action.
+    pub fn try_from_delegate_action(
+        delegate_action: &near_primitives::delegate_action::DelegateAction,
+        metadata: ActionMetadata,
+    ) -> Result<Vec<Self>, &'static str> {
+        let mut actions = Vec::with_capacity(delegate_action.actions.len());
+
+        for nearcore_action in delegate_action.clone().actions {
+            let action = match views::ActionView::from(
+                <near_primitives::delegate_action::NonDelegateAction as Into<
+                    near_primitives::transaction::Action,
+                >>::into(nearcore_action),
+            ) {
+                views::ActionView::CreateAccount => Self::CreateAccount(CreateAccount {
+                    metadata: metadata.clone(),
```

Comment:
@frol I wonder whether this metadata is related. Having the `DelegateAction` structure I believe the sender, receiver, and public key in the `ActionMetadata` for all nested `Actions` should be based on the fields from the `DelegateAction` itself rather than from the parent Transaction/Receipt, what do you think?

```rust
pub struct DelegateAction {
    /// Signer of the delegated actions
    pub sender_id: AccountId,
    /// Receiver of the delegated actions.
    pub receiver_id: AccountId,
    /// List of actions to be executed.
    ///
    /// With the meta transactions MVP defined in NEP-366, nested
    /// DelegateActions are not allowed. A separate type is used to enforce it.
    pub actions: Vec<NonDelegateAction>,
    /// Nonce to ensure that the same delegate action is not sent twice by a
    /// relayer and should match for given account's `public_key`.
    /// After this action is processed it will increment.
    pub nonce: Nonce,
    /// The maximal height of the block in the blockchain below which the given DelegateAction is valid.
    pub max_block_height: BlockHeight,
    /// Public key used to sign this delegated action.
    pub public_key: PublicKey,
}
```